src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -299,10 +299,236 @@ ss::future<> snc_quota_manager::quota_balancer_step() {
+        if (share.rem > 0) {
+            v += 1;
+            share.quot -= 1;
+        } else if (share.rem < 0) {
+            v -= 1;
+            share.quot += 1;
+        }
+    }
+}
+
+/// If \p value is less than \p limit, set it to the \p limit value and
+/// return the difference as positive number. Otherwise \p value is unchanged
+/// and 0 is returned
+quota_t cap_to_ceiling(quota_t& value, const quota_t limit) {
+    if (const quota_t d = limit - value; d > 0) {
+        value += d;
+        return d;
+    }
+    return 0;
+}
+
+/// If \p delta is negative, splits its value accourding to \p surplus values:
+/// - amount under total surplus is split pro rata to \p surplus amounts
+/// - amount over total surplus is split equally. Values in \p schedule
+/// are added to.
+/// \pre schedule.size() == surplus.size()
+void dispense_negative_deltas(
+  std::vector<quota_t>& schedule,
+  quota_t& delta,
+  const std::vector<quota_t>& surplus) {
+    if (delta >= 0) {
+        return;
+    }
+
+    const quota_t total_surplus = std::reduce(
+      surplus.cbegin(), surplus.cend(), quota_t{0}, std::plus{});
+
+    if (delta > -total_surplus) {
+        quota_t remainder = delta;
+        // pro rata to surpluses
+        for (size_t k = 0; k != ss::smp::count; ++k) {
+            const quota_t share = muldiv(delta, surplus.at(k), total_surplus);
+            schedule.at(k) += share;
+            remainder -= share;
+        }
+        // the remainder equally
+        if (remainder > 0) {
+            vlog(
+              klog.error,
+              "qb - Expected {} <= 0; delta: {}, total_surplus: {}, surpluses: "
+              "{}",
+              remainder,
+              delta,
+              total_surplus,
+              surplus);
+            return;
+        }
+        if (remainder < 0) {
+            const quota_t d = (remainder + 1)
+                                / static_cast<quota_t>(schedule.size())
+                              - 1;
+            for (quota_t& s : schedule) {
+                const quota_t dd = std::max(d, remainder);
+                remainder -= dd;
+                s += dd;
+            }
+        }
+
+    } else { // delta <= -total_surplus
+
+        // all surpluses
+        for (size_t k = 0; k != ss::smp::count; ++k) {
+            const quota_t share = -surplus.at(k);
+            schedule.at(k) += share;
+            delta -= share;
+        }
+        // the rest equally
+        auto share = std::div(delta, ss::smp::count);
+        for (quota_t& s : schedule) {
+            s += share.quot;
+            if (share.rem < 0) {
+                s += -1;
+                share.quot -= -1;
+            }
+        }
+    }
+}
+
+} // namespace
+
 ss::future<> snc_quota_manager::quota_balancer_update(
-  ingress_egress_state<dispense_quota_amount> /*shard_quotas_update*/) {
+  ingress_egress_state<dispense_quota_amount> shard_quotas_update) {
+    const auto units = co_await _balancer_mx.get_units();
```

Comment:
This could indefinitely wait if theres some sort of bug introduced, maybe use the version with the timeout, and have the caller retry on detection of exception? This way at least the behavior can be logged and identified

Reply:
There are two scopes waiting on this mutex, each running in a fiber different from the other's, never calling each other from inside the scope even asynchronously, and not waiting on anything except for the futures for cross-shard calls that also follow the aforementioned. I see no way to deadlock this. 

Still I see some space for being cautious, like with any syncronization. I have an async task to re-evaluate the need for the 2nd fiber with Noah; if the other fiber and the mutex stay in the code after that, I will add timeouts on `get_units()` as you propose. Does that sounds ok?

Reply:
I can't identify any specific scenario that could lead to a deadlock on the mutex. Generally I see the pattern:

1. take mutex
2. perform a bounded amount of work
3. release mutex

I think in order for this to be a concern we need to identify a specific scneario. @graphcareful afaict if any bug is introduce that would do something like throw an exception then the mutex is automatically released. it's unusable at that point, but at least it doesn't deadlock.

i think that @graphcareful might be referring to a specific design pattern in which "things fail" and we bubble this up to a point where it is convenient to reset the state of the system to a known state. i wouldn't be opposed to that, but i'm also not seeing it as a big concern here since the cross-core operations are 100% cpu calculation--very simple and benign.